### DISCUSSION TITLE
Using urql subscriptions with Django/Python backend

### DISCUSSION THREAD

**Discussion by nhill-cpi** [urql-graphql/urql#1206] [Q&A] [answered]
Hi all,

I'm reading the [url subscriptions documentation](https://formidable.com/open-source/urql/docs/advanced/subscriptions/) and the two examples they give of subscription clients use Apollo Server and `graphqh-ws`. Do you know how I might set up a subscription for a Django/python backend? Would I pass in a websocket client to the `subscriptionExchanges`? Thanks!

**Comment by JoviDeCroock** [comment, marked as answer]
The backend implementation doesn't really matter, if you expose a `ws:endpoint` it will connect when you use either of these client-libraries. So both `graphql-ws` and `subscriptions-transport-ws` will work on the client, you need to pass the endpoint that is exposed from your backend

**Reply by nhill-cpi** [reply]
Thanks for your quick reply @JoviDeCroock . So even though I'm using the `urql` client, I would use the `graphql-ws` client just for the `urql` subscriptions part? Sorry if I'm misunderstanding something.

**Reply by JoviDeCroock** [reply]
Yes so for instance, your django GraphQL backend exposes a subscriptions endpoint on `ws://localhost:3000` then we can connect from our client-side urql like this:

```
import { Client, defaultExchanges, subscriptionExchange } from 'urql';
import { SubscriptionClient } from 'subscriptions-transport-ws';

const subscriptionClient = new SubscriptionClient('ws://localhost:3000/graphql', { reconnect: true }); // the websocket endpoint

const client = new Client({
  url: 'http://localhost:3000/graphql', // The regular query and mutations endpoint
  exchanges: [
    ...defaultExchanges,
    subscriptionExchange({
      forwardSubscription(operation) {
        return subscriptionClient.request(operation);
      },
    }),
  ],
});
```

We don't offer any utilities for the server-side (your DJango stuff) we handle the client-side, which is a protocol that will be followed by your DJango GraphQL implementation

**Reply by nhill-cpi** [reply]
Great, thanks for the clarification. I've got the backend covered - just wondered about the subscription client.
